File: python/app/schemas/order.py
"""
Order schemas for request/response validation
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class OrderBase(BaseModel):
    """Base order schema"""
    order_no: str = Field(..., description="订单编号")
    user_id: int = Field(..., description="用户ID")
    user_name: str = Field(..., description="用户姓名")
    user_phone: str = Field(..., description="用户手机号")
    user_id_card: Optional[str] = Field(None, description="用户身份证")
    user_email: Optional[str] = Field(None, description="用户邮箱")
    user_address: Optional[str] = Field(None, description="用户地址")
    product_id: int = Field(..., description="商品ID")
    product_name: str = Field(..., description="商品名称")
    product_category: str = Field(..., description="商品分类")
    product_price: Decimal = Field(..., description="商品单价")
    quantity: int = Field(..., description="购买数量")
    total_amount: Decimal = Field(..., description="订单总金额")
    discount_amount: Decimal = Field(default=Decimal("0.00"), description="优惠金额")
    pay_amount: Decimal = Field(..., description="实付金额")
    order_status: str = Field(..., description="订单状态")
    payment_method: Optional[str] = Field(None, description="支付方式")
    payment_time: Optional[datetime] = Field(None, description="支付时间")
    order_source: Optional[str] = Field(None, description="订单来源")
    shipping_address: Optional[str] = Field(None, description="收货地址")
    receiver_name: Optional[str] = Field(None, description="收货人姓名")
    receiver_phone: Optional[str] = Field(None, description="收货人电话")
    logistics_no: Optional[str] = Field(None, description="物流单号")
    delivery_time: Optional[datetime] = Field(None, description="发货时间")
    complete_time: Optional[datetime] = Field(None, description="完成时间")
    remark: Optional[str] = Field(None, description="备注")


class OrderResponse(OrderBase):
    """Order response schema"""
    order_id: int = Field(..., description="订单ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    is_deleted: int = Field(default=0, description="是否删除")

    class Config:
        from_attributes = True


class OrderQueryParams(BaseModel):
    """Order query parameters"""
    page: int = Field(default=1, ge=1, description="页码")
    page_size: int = Field(default=20, ge=1, le=1000, description="每页数量")
    order_no: Optional[str] = Field(None, description="订单编号")
    user_id: Optional[int] = Field(None, description="用户ID")
    order_status: Optional[str] = Field(None, description="订单状态")
    start_time: Optional[datetime] = Field(None, description="开始时间")
    end_time: Optional[datetime] = Field(None, description="结束时间")
    min_amount: Optional[Decimal] = Field(None, ge=0, description="最小金额")
    max_amount: Optional[Decimal] = Field(None, ge=0, description="最大金额")

    class Config:
        json_schema_extra = {
            "example": {
                "page": 1,
                "page_size": 20,
                "order_no": "ORD202401010001",
                "user_id": 10001,
                "order_status": "completed",
                "start_time": "2024-01-01T00:00:00",
                "end_time": "2024-12-31T23:59:59",
                "min_amount": 100.00,
                "max_amount": 10000.00,
            }
        }


class OrderListResponse(BaseModel):
    """Order list response with pagination"""
    total: int = Field(..., description="总记录数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")
    items: list[OrderResponse] = Field(..., description="订单列表")

    class Config:
        json_schema_extra = {
            "example": {
                "total": 100,
                "page": 1,
                "page_size": 20,
                "total_pages": 5,
                "items": [],
            }
        }

File: golang/internal/model/order.go
package model

import (
	"time"
)

type Order struct {
	OrderID         int64          `gorm:"primaryKey;autoIncrement" json:"order_id"`
	OrderNo         string         `gorm:"type:varchar(32);uniqueIndex;not null" json:"order_no"`
	UserID          int64          `gorm:"not null;index" json:"user_id"`
	UserName        string         `gorm:"type:varchar(64)" json:"user_name"`
	UserPhone       string         `gorm:"type:varchar(11)" json:"user_phone"`
	UserIDCard      string         `gorm:"type:varchar(18)" json:"user_id_card"`
	UserEmail       string         `gorm:"type:varchar(128)" json:"user_email"`
	UserAddress     string         `gorm:"type:varchar(256)" json:"user_address"`
	ProductID       int64          `gorm:"not null" json:"product_id"`
	ProductName     string         `gorm:"type:varchar(128)" json:"product_name"`
	ProductCategory string         `gorm:"type:varchar(64)" json:"product_category"`
	ProductPrice    float64        `gorm:"type:decimal(10,2);not null" json:"product_price"`
	Quantity        int            `gorm:"not null" json:"quantity"`
	TotalAmount     float64        `gorm:"type:decimal(12,2);not null;index" json:"total_amount"`
	DiscountAmount  float64        `gorm:"type:decimal(12,2);default:0" json:"discount_amount"`
	PayAmount       float64        `gorm:"type:decimal(12,2);not null" json:"pay_amount"`
	OrderStatus     string         `gorm:"type:varchar(16);not null;default:'pending';index" json:"order_status"`
	PaymentMethod   string         `gorm:"type:varchar(16)" json:"payment_method"`
	PaymentTime     *time.Time     `json:"payment_time"`
	OrderSource     string         `gorm:"type:varchar(32)" json:"order_source"`
	ShippingAddress string         `gorm:"type:varchar(512)" json:"shipping_address"`
	ReceiverName    string         `gorm:"type:varchar(64)" json:"receiver_name"`
	ReceiverPhone   string         `gorm:"type:varchar(11)" json:"receiver_phone"`
	LogisticsNo     string         `gorm:"type:varchar(32)" json:"logistics_no"`
	DeliveryTime    *time.Time     `json:"delivery_time"`
	CompleteTime    *time.Time     `json:"complete_time"`
	Remark          string         `gorm:"type:varchar(512)" json:"remark"`
	CreatedAt       time.Time      `gorm:"not null;default:CURRENT_TIMESTAMP;index" json:"created_at"`
	UpdatedAt       time.Time      `gorm:"not null;default:CURRENT_TIMESTAMP" json:"updated_at"`
	IsDeleted       int16          `gorm:"not null;default:0" json:"is_deleted"`
}

func (Order) TableName() string {
	return "orders"
}

type OrderQueryParams struct {
	Page       int        `form:"page" json:"page"`
	PageSize   int        `form:"page_size" json:"page_size"`
	StartTime  *time.Time `form:"start_time" json:"start_time" time_format:"2006-01-02T15:04:05Z07:00"`
	EndTime    *time.Time `form:"end_time" json:"end_time" time_format:"2006-01-02T15:04:05Z07:00"`
	OrderNo    string     `form:"order_no" json:"order_no"`
	UserID     *int64     `form:"user_id" json:"user_id"`
	Status     string     `form:"status" json:"status"`
	MinAmount  *float64   `form:"min_amount" json:"min_amount"`
	MaxAmount  *float64   `form:"max_amount" json:"max_amount"`
	SortBy     string     `form:"sort_by" json:"sort_by"`
	SortOrder  string     `form:"sort_order" json:"sort_order"`
}

func (p *OrderQueryParams) GetOffset() int {
	if p.Page <= 0 {
		p.Page = 1
	}
	if p.PageSize <= 0 {
		p.PageSize = 20
	}
	if p.PageSize > 1000 {
		p.PageSize = 1000
	}
	return (p.Page - 1) * p.PageSize
}

func (p *OrderQueryParams) GetLimit() int {
	if p.PageSize <= 0 {
		p.PageSize = 20
	}
	if p.PageSize > 1000 {
		p.PageSize = 1000
	}
	return p.PageSize
}

type OrderListResponse struct {
	Total int64   `json:"total"`
	Page  int     `json:"page"`
	Size  int     `json:"size"`
	Items []Order `json:"items"`
}
